code/TagField.php

```diff
@@ -173,4 +274,123 @@ public function saveInto(DataObjectInterface $record) {
 			$record->$name = implode(',', $values);
 		}
 	}
+
+	/**
+	 * Returns a JSON string of tags, for ajax-based search.
+	 *
+	 * @param SS_HTTPRequest $request
+	 *
+	 * @return SS_HTTPResponse
+	 */
+	public function suggest(SS_HTTPRequest $request) {
+		$recordClass = $this->getRecordClass();
+
+		if($recordClass === null) {
+			return '[]';
+		}
+
+		$name = $this->getName();
+
+		/**
+		 * @var DataObject $object
+		 */
+		$object = singleton($recordClass);
+
+		$term = $request->getVar('term');
+
+		$tags = array();
+
+		if($object->hasMethod($name)) {
+			$tags = $this->getObjectTags($object, $term);
+		} elseif($object->hasField($name)) {
+			$tags = $this->getStringTags($term);
+		}
+
+		$response = new SS_HTTPResponse();
+
+		$response->setBody(Convert::raw2json(
+			array('items' => $tags)
+		));
+
+		$response->addHeader('Content-Type', 'application/json');
+
+		return $response;
+	}
+
+	/**
+	 * Returns array of arrays representing DataObject-based tags.
+	 *
+	 * @param DataObject $instance
+	 * @param string $term
+	 *
+	 * @return array
+	 */
+	protected function getObjectTags(DataObject $instance, $term) {
+		$name = $this->getName();
+		$relationTitle = $this->getRelationTitle();
+
+		$relation = $instance->{$name}();
+
+		$term = Convert::raw2sql($term);
+
+		$query = DataList::create($relation->dataClass())
+			->where('"' . $relationTitle . '" LIKE \'%' . $term . '%\'')
+			->sort($relationTitle)
+			->limit($this->getAjaxItemLimit());
+
+		$items = array();
+
+		foreach($query as $record) {
+			$text = $record->{$relationTitle};
+
+			if (!in_array($text, $items)) {
+				$items[] = array(
+					'id' => $record->ID,
+					'text' => $record->{$relationTitle}
+				);
+			}
+		}
+
+		return $items;
+	}
+
+	/**
+	 * Returns array of arrays representing string-based tags.
+	 *
+	 * @param string $term
+	 *
+	 * @return array
+	 */
+	protected function getStringTags($term) {
+		$name = $this->getName();
+		$recordClass = $this->getRecordClass();
+
+		$term = Convert::raw2sql($term);
+
+		$column = '"' . $recordClass . '"."' . $name . '"';
+		$value = '\'%' . $term . '%\'';
+
+		$query = new DataQuery($recordClass);
+
+		$query
+			->where('LOWER(' . $column . ') LIKE LOWER(' . $value . ')')
```

Comment:
Same as above; Better to use filter api and ->column or ->map to only select the necessary fields.